src/memory.rs

```diff
@@ -0,0 +1,498 @@
+                get_optional_word(&mut mem_pointer, &structure.data, 0);
+        }
+        if structure.version > (2, 8).into() {
+            memory.minimum_voltage = get_optional_word(&mut mem_pointer, &structure.data, 0);
+            memory.maximum_voltage = get_optional_word(&mut mem_pointer, &structure.data, 0);
+            memory.configured_voltage = get_optional_word(&mut mem_pointer, &structure.data, 0);
+        }
+        if structure.version > (3, 2).into() {
+            memory.memory_technology = Some(MemoryTechnology::from(structure.data[mem_pointer]));
+            mem_pointer += 1;
+            memory.operating_mode_capability = Some(OperatingModes::from_bits_truncate(get_word(
+                &mut mem_pointer,
+                &structure.data,
+            )));
+            memory.firmware_version = Some(find_string(&structure, &mut mem_pointer)?);
+            memory.module_manufacturer = Some(get_word(&mut mem_pointer, &structure.data));
+            memory.module_product_id = Some(get_word(&mut mem_pointer, &structure.data));
+            memory.memory_subsystem_controller_manufacturer_id =
+                Some(get_word(&mut mem_pointer, &structure.data));
+            memory.memory_subsystem_controller_product_id =
+                Some(get_word(&mut mem_pointer, &structure.data));
+            memory.non_volatile_size =
+                get_optional_qword(&mut mem_pointer, &structure.data, 0xFFFFFFFFFFFFFFFF);
+            memory.volatile_size =
+                get_optional_qword(&mut mem_pointer, &structure.data, 0xFFFFFFFFFFFFFFFF);
+            memory.cache_size =
+                get_optional_qword(&mut mem_pointer, &structure.data, 0xFFFFFFFFFFFFFFFF);
+            memory.logical_size =
+                get_optional_qword(&mut mem_pointer, &structure.data, 0xFFFFFFFFFFFFFFFF);
+        }
+        if structure.version > (3, 3).into() {
+            memory.extended_speed = Some(get_dword(&mut mem_pointer, &structure.data));
+            memory.extended_configured_memory_speed =
+                Some(get_dword(&mut mem_pointer, &structure.data));
+        }
+
+        Ok(memory)
+    }
+}
+
+fn find_string<'buffer>(
+    structure: &super::RawStructure<'buffer>,
+    pointer: &mut usize,
+) -> Result<&'buffer str, super::MalformedStructureError> {
+    let s = structure.find_string(structure.data[*pointer])?;
+    *pointer += 1;
+    Ok(s)
+}
+
+fn get_optional_qword(pointer: &mut usize, data: &[u8], none_val: u64) -> Option<u64> {
+    let word = get_qword(pointer, data);
+    if word == none_val {
+        None
+    } else {
+        Some(word)
+    }
+}
+
+fn get_optional_word(pointer: &mut usize, data: &[u8], none_val: u16) -> Option<u16> {
+    let word = get_word(pointer, data);
+    if word == none_val {
+        None
+    } else {
+        Some(word)
+    }
+}
+
+fn get_word(pointer: &mut usize, data: &[u8]) -> u16 {
+    let word = u16::from_le_bytes(data[*pointer..(*pointer + 2)].try_into().unwrap());
```

Comment:
I'm not familiar with how the fail crate works so I used unwrap here but I'd like to remove that if possible

Reply:
I'm going to remove these unwrap's so hang tight. I'll push another commit in a few mins

Reply:
Will do.